File: app/src/main/java/com/ihomey/linkuphome/data/repository/NetworkBoundResource.java
package com.ihomey.linkuphome.data.repository;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;

import com.ihomey.linkuphome.AppExecutors;
import com.ihomey.linkuphome.data.vo.Resource;

/**
 * Created by dongcaizheng on 2018/1/11.
 */

public abstract class NetworkBoundResource<ResultType> {

    private final MediatorLiveData<Resource<ResultType>> result = new MediatorLiveData<>();

    @MainThread
    NetworkBoundResource(AppExecutors appExecutors) {
        result.setValue(Resource.loading(null));
        appExecutors.diskIO().execute(() -> {
            LiveData<ResultType> dbSource = loadFromDb();
            appExecutors.mainThread().execute(() ->
                    result.addSource(dbSource, value -> result.setValue(Resource.success(value)))
            );
        });
    }


    public LiveData<Resource<ResultType>> asLiveData() {
        return result;
    }

    @NonNull
    @MainThread
    protected abstract LiveData<ResultType> loadFromDb();

}

File: app/src/main/java/com/ihomey/linkuphome/data/api/NetworkBoundResource.kt
package com.ihomey.linkuphome.data.api

import android.text.TextUtils
import android.util.Log
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import com.ihomey.linkuphome.AppExecutors
import com.ihomey.linkuphome.data.vo.Resource


/**
 * A generic class that can provide a resource backed by both the sqlite database and the network.
 *
 *
 * You can read more about it in the [Architecture
 * Guide](https://developer.android.com/arch).
 * @param <ResultType>
 * @param <RequestType>
</RequestType></ResultType> */
abstract class NetworkBoundResource<ResultType>
@MainThread constructor(private val appExecutors: AppExecutors) {

    private val result = MediatorLiveData<Resource<ResultType>>()

    init {
        result.value = Resource.loading(null)
        @Suppress("LeakingThis")
        val dbSource = loadFromDb()
        result.addSource(dbSource) { data ->
            result.removeSource(dbSource)
            if (shouldFetch(data)) {
                fetchFromNetwork(dbSource)
            } else {
                result.addSource(dbSource) { newData ->
                    setValue(Resource.success(newData))
                }
            }
        }
    }

    @MainThread
    private fun setValue(newValue: Resource<ResultType>) {
        if (result.value != newValue) {
            result.value = newValue
        }
    }

    private fun fetchFromNetwork(dbSource: LiveData<ResultType>) {
        val apiResponse = createCall()
        // we re-attach dbSource as a new source, it will dispatch its latest value quickly
        result.addSource(dbSource) { newData ->
            setValue(Resource.loading(newData))
        }
        result.addSource(apiResponse) { response ->
            result.removeSource(dbSource)
            if (TextUtils.equals("0000",response?.code)) {
                appExecutors.diskIO().execute {
                    saveCallResult(processResponse(response.data))
                    appExecutors.mainThread().execute {
                        setValue(Resource.success(response.data))
                    }
                }
            } else {
                onFetchFailed()
                result.addSource(dbSource) { newData ->
                    setValue(Resource.error(response?.code, newData))
                }
            }
        }
    }

    protected open fun onFetchFailed() {}

    fun asLiveData() = result as LiveData<Resource<ResultType>>

    @WorkerThread
    protected open fun processResponse(response: ResultType?) = response

    @WorkerThread
    protected abstract fun saveCallResult(item: ResultType?)

    @MainThread
    protected abstract fun shouldFetch(data: ResultType?): Boolean

    @MainThread
    protected abstract fun loadFromDb(): LiveData<ResultType>

    @MainThread
    protected abstract fun createCall(): LiveData<ApiResult<ResultType>>
}